### DISCUSSION TITLE
Using featureContext and ScenarioContext with Allure Specflow

### DISCUSSION THREAD

**Discussion by Divya23490** [allure-framework/allure2#2191] [General Discussion]
I am working on Allure reporting. Using Allure.Specflow version 2.9.5-preview. I am also using the FeatureContext and ScenarioContext using ContextInjection. While running the before Scenario inside Hooks I get one error 
System.InvalidCastException : Specified cast is not valid.
TearDown : System.NullReferenceException : Object reference not set to an instance of an object
  at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.FireEvents (TechTalk.SpecFlow.Bindings.HookType hookType) [0x0013f] in <d57596056d214977bd041e46fd77023a>:0 
  at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.FireScenarioEvents (TechTalk.SpecFlow.Bindings.HookType bindingEvent) [0x00000] in <d57596056d214977bd041e46fd77023a>:0 
  at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.OnScenarioStart () [0x00037] in <d57596056d214977bd041e46fd77023a>:0 
  at TechTalk.SpecFlow.Infrastructure.TestExecutionEngine.OnAfterLastStep () [0x0020f] in <d57596056d214977bd041e46fd77023a>:0 
  at TechTalk.SpecFlow.TestRunner.CollectScenarioErrors () [0x00000] in <d57596056d214977bd041e46fd77023a>:0 
  
  Can someone help me understand the issue and how can I fix it.

**Comment by delatrie** [comment]
Could be anything. Share some code, please.

**Reply by Divya23490** [reply]
@delatrie , added the code for your reference. Can you share some examples in which the Hooks attribute uses FeatureContext and ScenarioContext. I also checked the SelectedTestRun as I running Specflow scenario inside one of the feature file. I have updated the project to .net 7 and using the SpecFlow.NUnit with 3.9.52 and SpecFlow.Tools.MsBuild.Generation with the same version. I have also added the Lib.Harmony with 2.3.0-prerelease.2. I have updated the Allure.Specflow package to 2.10. But now I am getting error in BeforeFeature attribute
OneTimeSetUp: System.InvalidOperationException : Unable to change the container context because the fixture context is active.
  Exception doesn't have a stacktrace
  This is the code 
      public static void BeforeFeature(FeatureContext featureContext)
    {
      
        var featureTitle = featureContext.FeatureInfo.Title;
        var featureId = Guid.NewGuid().ToString();
       
        // Check if the feature failed
        if (TestContext.CurrentContext.Result.Outcome.Status != TestStatus.Failed && !string.IsNullOrEmpty(featureTitle))
        {
            if (_isFixtureActive)
            {
                _allureLifecycle.StopFixture();
                _isFixtureActive = false;
            }

            if (!_isFixtureActive)
            {
                var testContainer = new TestResultContainer()
                {
                    uuid = featureId,
                    name = featureTitle
                };
                _allureLifecycle.StartTestContainer(testContainer);
            }
            
        }
    
        // Reset all test retry variables
        featureContext.Add("featureId",featureId);
        featureContext.Add("ScenarioCount", 0);
        featureContext.Add("FailCount", 0);
        featureContext.Add("StepFail", 0);
    }
     [BeforeFeature]
        public static void BeforeFeature(FeatureContext featureContext)
 
            // Temp local report handling
            AllureReporting.BeforeFeature(featureContext);
        }

**Comment by Divya23490** [comment]
I am working on creating the custom Allure Adapter 
This is the code inside Hooks class 
[BeforeScenario]
public static void BeforeScenario(FeatureContext featureContext,ScenarioContext scenarioContext)
          {
            // Temp local report handling
            AllureReporting.Initialize(featureContext,scenarioContext);
            }
            
 This is the Initialize method in Allure Reporting. I am getting failure after completion of the method. 
  public static void Initialize(FeatureContext featureContext, ScenarioContext scenarioContext)
    {
        try
        {
            // Get the current scenario title
            var scenarioTitle = scenarioContext.ScenarioInfo.Title;

            // Get the feature ID from FeatureContext
            var featureId = featureContext.Get<string>("featureId");

            // Check if featureId is not null
            if (featureId != null)
            {
                // Create a new testcase with a UUID and add it to the test container
                var testCase = new TestResult()
                {
                    uuid = Guid.NewGuid().ToString(),
                    name = scenarioTitle
                };

                // Start the test case
                _allureLifecycle.StartTestCase(featureId, testCase);

                // Set new value of the Scenario counter to Feature context
                scenarioContext.Add("StepCount", 1);
                featureContext["ScenarioCount"] = featureContext.Get<int>("ScenarioCount") + 1;
                featureContext["StepFail"] = 0;

                // Add attachments or additional setup if needed
            }
            else
            {
                Console.WriteLine("Feature ID is null. Skipping Allure reporting setup for this scenario.");
            }
        }
        catch (Exception ex)
        {
            Console.WriteLine("Error occurred during Allure reporting initialization: " + ex.Message);
        }
    }
    I get the error inside the BeforeScenario attribute.
